File: frontend/src/utils/aqi.js
// US EPA composite AQI. Mirrors backend/aqi.py - keep the two in sync.
export const AQI_BREAKPOINTS = {
  pm25: [[0.0, 12.0, 0, 50], [12.1, 35.4, 51, 100], [35.5, 55.4, 101, 150], [55.5, 150.4, 151, 200], [150.5, 250.4, 201, 300], [250.5, 500.4, 301, 500]],
  pm10: [[0, 54, 0, 50], [55, 154, 51, 100], [155, 254, 101, 150], [255, 354, 151, 200], [355, 424, 201, 300], [425, 604, 301, 500]],
  co: [[0.0, 4.4, 0, 50], [4.5, 9.4, 51, 100], [9.5, 12.4, 101, 150], [12.5, 15.4, 151, 200], [15.5, 30.4, 201, 300], [30.5, 50.4, 301, 500]], // ppm
  no2: [[0, 53, 0, 50], [54, 100, 51, 100], [101, 360, 101, 150], [361, 649, 151, 200], [650, 1249, 201, 300], [1250, 2049, 301, 500]], // ppb
  o3: [[0.0, 0.054, 0, 50], [0.055, 0.07, 51, 100], [0.071, 0.085, 101, 150], [0.086, 0.105, 151, 200], [0.106, 0.2, 201, 300]], // ppm, 8-hr table
};

function subAqi(conc, table) {
  if (conc == null || Number.isNaN(conc)) return null;
  for (const [cLow, cHigh, aLow, aHigh] of table) {
    if (conc >= cLow && conc <= cHigh) {
      return Math.round(((aHigh - aLow) / (cHigh - cLow)) * (conc - cLow) + aLow);
    }
  }
  const last = table[table.length - 1];
  return conc > last[1] ? 500 : 0;
}

export function computeCompositeAQI(aq) {
  const candidates = [
    { name: 'PM2.5', aqi: subAqi(aq.pm2_5, AQI_BREAKPOINTS.pm25) },
    { name: 'PM10', aqi: subAqi(aq.pm10, AQI_BREAKPOINTS.pm10) },
    { name: 'CO', aqi: subAqi(aq.carbon_monoxide / 1145, AQI_BREAKPOINTS.co) },
    { name: 'NO₂', aqi: subAqi(aq.nitrogen_dioxide / 1.88, AQI_BREAKPOINTS.no2) },
    { name: 'O₃', aqi: subAqi(aq.ozone / 1960, AQI_BREAKPOINTS.o3) },
  ].filter((c) => c.aqi != null);
  if (!candidates.length) return null;
  candidates.sort((a, b) => b.aqi - a.aqi);
  return { aqi: candidates[0].aqi, dominant: candidates[0].name };
}

export function aqiCategory(aqi) {
  if (aqi <= 50) return { label: 'Good', color: '#22c55e' };
  if (aqi <= 100) return { label: 'Moderate', color: '#eab308' };
  if (aqi <= 150) return { label: 'Unhealthy for Sensitive Groups', color: '#f97316' };
  if (aqi <= 200) return { label: 'Unhealthy', color: '#ef4444' };
  if (aqi <= 300) return { label: 'Very Unhealthy', color: '#a855f7' };
  return { label: 'Hazardous', color: '#7f1d1d' };
}

export function classifyLadder(value, ladder) {
  for (const step of ladder) {
    if (value <= step.max) return step;
  }
  return ladder[ladder.length - 1];
}

export const POLLUTANT_LADDERS = {
  pm25: [{ max: 12.0, cls: 'success', label: 'Good' }, { max: 35.4, cls: 'warning', label: 'Moderate' }, { max: Infinity, cls: 'error', label: 'Unhealthy' }],
  pm10: [{ max: 54, cls: 'success', label: 'Good' }, { max: 154, cls: 'warning', label: 'Moderate' }, { max: Infinity, cls: 'error', label: 'Unhealthy' }],
  no2: [{ max: 100, cls: 'success', label: 'Good' }, { max: 188, cls: 'warning', label: 'Moderate' }, { max: Infinity, cls: 'error', label: 'Unhealthy' }],
  o3: [{ max: 108, cls: 'success', label: 'Good' }, { max: 137, cls: 'warning', label: 'Moderate' }, { max: Infinity, cls: 'error', label: 'Unhealthy' }],
  co: [{ max: 5038, cls: 'success', label: 'Good' }, { max: 10763, cls: 'warning', label: 'Moderate' }, { max: Infinity, cls: 'error', label: 'Unhealthy' }],
};

export function classifyTemp(temp) {
  if (temp >= 18 && temp <= 30) return { cls: 'success', label: 'Comfortable' };
  if ((temp >= 10 && temp < 18) || (temp > 30 && temp <= 35)) return { cls: 'warning', label: 'Caution' };
  return { cls: 'error', label: 'Extreme' };
}

export function classifyHumidity(h) {
  if (h >= 30 && h <= 60) return { cls: 'success', label: 'Comfortable' };
  if ((h >= 20 && h < 30) || (h > 60 && h <= 75)) return { cls: 'warning', label: 'Humid' };
  return { cls: 'error', label: 'Extreme' };
}

File: backend/aqi.py
"""
US EPA composite AQI calculation, mirrored from the JS implementation in
frontend/index.html (computeCompositeAQI/subAQI). Keep the two in sync -
there's no shared config file between the Python backend and the vanilla-JS
frontend in this project.
"""

# (cLow, cHigh, aLow, aHigh) breakpoints per pollutant.
AQI_BREAKPOINTS = {
    "pm2_5": [
        (0.0, 12.0, 0, 50), (12.1, 35.4, 51, 100), (35.5, 55.4, 101, 150),
        (55.5, 150.4, 151, 200), (150.5, 250.4, 201, 300), (250.5, 500.4, 301, 500),
    ],
    "pm10": [
        (0, 54, 0, 50), (55, 154, 51, 100), (155, 254, 101, 150),
        (255, 354, 151, 200), (355, 424, 201, 300), (425, 604, 301, 500),
    ],
    "carbon_monoxide": [  # ppm
        (0.0, 4.4, 0, 50), (4.5, 9.4, 51, 100), (9.5, 12.4, 101, 150),
        (12.5, 15.4, 151, 200), (15.5, 30.4, 201, 300), (30.5, 50.4, 301, 500),
    ],
    "nitrogen_dioxide": [  # ppb
        (0, 53, 0, 50), (54, 100, 51, 100), (101, 360, 101, 150),
        (361, 649, 151, 200), (650, 1249, 201, 300), (1250, 2049, 301, 500),
    ],
    "ozone": [  # ppm, 8-hr table
        (0.000, 0.054, 0, 50), (0.055, 0.070, 51, 100), (0.071, 0.085, 101, 150),
        (0.086, 0.105, 151, 200), (0.106, 0.200, 201, 300),
    ],
}


def _sub_aqi(conc, table):
    if conc is None:
        return None
    for c_low, c_high, a_low, a_high in table:
        if c_low <= conc <= c_high:
            return round(((a_high - a_low) / (c_high - c_low)) * (conc - c_low) + a_low)
    return 500 if conc > table[-1][1] else 0


def composite_aqi(reading: dict):
    """reading: dict with pm2_5/pm10/carbon_monoxide/nitrogen_dioxide/ozone in
    ug/m3 (matching the readings table / Open-Meteo field names). Returns the
    AQI number (int) or None if no pollutant values are present."""
    candidates = []
    if reading.get("pm2_5") is not None:
        candidates.append(_sub_aqi(reading["pm2_5"], AQI_BREAKPOINTS["pm2_5"]))
    if reading.get("pm10") is not None:
        candidates.append(_sub_aqi(reading["pm10"], AQI_BREAKPOINTS["pm10"]))
    if reading.get("carbon_monoxide") is not None:
        candidates.append(_sub_aqi(reading["carbon_monoxide"] / 1145, AQI_BREAKPOINTS["carbon_monoxide"]))
    if reading.get("nitrogen_dioxide") is not None:
        candidates.append(_sub_aqi(reading["nitrogen_dioxide"] / 1.88, AQI_BREAKPOINTS["nitrogen_dioxide"]))
    if reading.get("ozone") is not None:
        candidates.append(_sub_aqi(reading["ozone"] / 1960, AQI_BREAKPOINTS["ozone"]))
    candidates = [c for c in candidates if c is not None]
    return max(candidates) if candidates else None
